pycloudlib/vmware/cloud.py

```diff
@@ -0,0 +1,256 @@
+# This file is part of pycloudlib. See LICENSE file for license information.
+"""VMWare Cloud type."""
+
+import base64
+import os
+import shutil
+import subprocess
+from itertools import count
+from typing import Optional
+
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.config import ConfigFile
+from pycloudlib.util import add_key_to_cloud_config
+from pycloudlib.vmware.instance import VMWareInstance
+
+SERIES_TO_TEMPLATE = {
+    "focal": "TEMPLATE-cloud-init-focal",
+    "jammy": "TEMPLATE-cloud-init-jammy",
+}
+
+METADATA = "instance-id: pycloudlib-vm"
+
+
+class VMWare(BaseCloud):
+    """VMWare Cloud class."""
+
+    _type = "vmware"
+    _instance_counter = count()
+
+    def __init__(
+        self,
+        tag: str,
+        timestamp_suffix: bool = True,
+        config_file: Optional[ConfigFile] = None,
+        *,
+        server=None,
+        username=None,
+        password=None,
+        datacenter=None,
+        datastore=None,
+        folder=None,
+        insecure_transport=None,
+    ):
+        """Initialize VMWare cloud class.
+
+        Args:
+            tag: string used to name and tag resources with
+            timestamp_suffix: Append a timestamped suffix to the tag string.
+            config_file: path to pycloudlib configuration file
+        """
+        super().__init__(
+            tag=tag,
+            timestamp_suffix=timestamp_suffix,
+            config_file=config_file,
+            required_values=[
+                server,
+                username,
+                password,
+                datacenter,
+                datastore,
+                folder,
+                insecure_transport,
+            ],
+        )
+        server = server or self.config.get("server")
+        username = username or self.config.get("username")
+        password = password or self.config.get("password")
+        datacenter = datacenter or self.config.get("datacenter")
+        datastore = datastore or self.config.get("datastore")
+        folder = folder or self.config.get("folder")
+        insecure_transport = insecure_transport or self.config.get(
+            "insecure_transport"
+        )
+
+        self.govc = shutil.which("govc")
+        if not self.govc:
+            raise ValueError(
+                "'govc' must be installed and added to PATH. See "
+                "https://github.com/vmware/govmomi/tree/main/govc#installation"
+            )
+        self.env = {
+            "GOVC_INSECURE": "1" if insecure_transport else "0",
+            "GOVC_URL": server,
+            "GOVC_USERNAME": username,
+            "GOVC_PASSWORD": password,
+            "GOVC_DATACENTER": datacenter,
+            "GOVC_DATASTORE": datastore,
+            "GOVC_FOLDER": folder,
+            "PATH": os.environ.get("PATH"),
```

Comment:
Again, we don't want to source these variable from the environment because we want to provide them via pycloudlib.toml right? This way pycloudlib.toml overrides any environment variables... and provides a single file with secrets which is removed across test runs right?

Reply:
Yes...not sure if this is just asking for clarification or asking for a change, but the `env` being set here is only the env used by specific subp calls. The default environment is not modified, nor can the default environment be used. We need an env set here because that's how `govc` gets the values it needs to interact with vSphere. Priority for these env vars is:
1. Any argument passed to the function
2. pycloudlib.toml

If either of those two aren't provided, we should get a KeyError with `<setting> must be defined in pycloudlib.toml to make this call`.